README.md
# MRDP: Monthly Report Data Processing

Overview
========

MRDP is a program to create routine geomagnetic monthly report at geomagnetic observatory.
It reads IAGA-2002 geomagnetic data format and write the report in Microsoft
Excel format. It is also can be used to remove step in magnetogram.


Requirements
===============================

Officially, MRDP requires Python version 3.7 and KASm installed on your computer.
Works better using Anaconda.
You can get KASm from <http://www.intermagnet.org/software/Kasm_1.09.zip>

Python module requirements :
- PyQt5
- Numpy
- Openpyxl
- Matplotlib
- Pillow
- Scipy
- Pandas
- Lxml


Instant running
===============================

You can run MRDP from this directory by clicking MRDP.bat directly or by creating desktop shortcut.
Adjust station.ini to change station parameters. 



CHANGES
0.0.1 (05-nov-16)
---------------
- First public release

0.0.2 (13-nov-16)
---------------
- Change output excel directory to C:\Users\USERNAME\data\QuasiDefinitive\Excel
- Change characters of K-index to numbers

0.0.3 (16-nov-16)
---------------
- Fix problem with progressbar

0.0.4 (17-nov-16)
---------------
- Fix problem when used with uncomplete data

0.0.5 (20-nov-16)
---------------
- Existing geomagnetic storm will not be replaced after executing program

1.0.0 (25-nov-16)
---------------
- Add signal and K Index plot

1.0.1 (25-jan-17)
---------------
- Add select excel output folder button

1.1.0 (20-mar-17)
---------------
- Change input, only reads IAGA-2002 files now

1.1.1 (16-jul-17)
---------------
- Fix data loss handling problem

1.1.2 (16-jul-17)
---------------
- Add station parameters as initiation file

2.0.0 (5-dec-17)
---------------
- Add step removal processing
- Fixed plot on K and A index

2.1.0 (10-dec-17)
---------------
- Add baseline plot
- Add splash screen
- Fix clipped problem when creating plots

2.2.0 (23-apr-20)
---------------
- Convert to Python 3
- Convert from PyQt4 to PyQt5

main.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 10 21:08:04 2020

@author: yosi setiawan
"""

import sys,os,glob,time
from calendar import monthrange
from os.path import expanduser
from PyQt5 import QtCore, QtGui, uic, QtWidgets
from process import iaga2imfv,min2hour,format_excel
from step_remove import step_removal

#QtWidgets.QApplication.setAttribute(QtCore.Qt.AA_EnableHighDpiScaling, True) #enable highdpi scaling
#QtWidgets.QApplication.setAttribute(QtCore.Qt.AA_UseHighDpiPixmaps, True) #use highdpi icons

# Set lokasi home folder
home = expanduser("~")
# Nama file GUI
qtCreatorFile = "gui.ui" # Enter file here.

# Load file GUI
Ui_MainWindow, QtBaseClass = uic.loadUiType(qtCreatorFile)


class EmittingStream(QtCore.QObject):
    textWritten = QtCore.pyqtSignal(str)
    def write(self, text):
        self.textWritten.emit(str(text))
        
    def flush(self):
        pass

class MyApp(QtWidgets.QMainWindow, Ui_MainWindow):
    def __init__(self):
        QtWidgets.QMainWindow.__init__(self)
        Ui_MainWindow.__init__(self)
        self.setupUi(self)
        self.SearchButtonIAGA.clicked.connect(self.SearchFolderIAGA)
        self.ExecuteButton.clicked.connect(self.Execute)
        self.StepRemovalButton.clicked.connect(self.stepremoval)
        self.CalendarWidget.clicked[QtCore.QDate].connect(self.showDate)
        self.progressBar.setMinimum(0)
        sys.stdout = EmittingStream(textWritten=self.normalOutputWritten)
        
    def __del__(self):
        # Restore sys.stdout
        sys.stdout = sys.__stdout__
        
    # Fungsi untuk tombol Search Folder IAGA
    def SearchFolderIAGA(self):
        global dirIAGA_,numfile
        dirIAGA_ = QtWidgets.QFileDialog.getExistingDirectory(None, 'Select a folder:', home+'\\data', QtWidgets.QFileDialog.ShowDirsOnly)
        self.labelfolderIAGA.setText(dirIAGA_)
        numfile = len([name for name in os.listdir(str(dirIAGA_)) if name.endswith('.min') and os.path.isfile(os.path.join(str(dirIAGA_), name))])
        return dirIAGA_,numfile
    
    # Fungsi untuk menampilkan teks
    def normalOutputWritten(self, text):
        cursor = self.textEdit.textCursor()
        cursor.movePosition(QtGui.QTextCursor.End)
        cursor.insertText(text)
        self.textEdit.setTextCursor(cursor)
        self.textEdit.ensureCursorVisible()
        
    # Fungsi untuk tombol Execute
    def Execute(self):
        listfile = glob.glob(os.path.join(str(dirIAGA_), '*.min'))
        # Menentukan file data pertama untuk diambil tahun bulan tanggalnya
        first_data = os.path.basename(listfile[0])
        tahun1 = int(first_data[3:7])
        bulan1 = int(first_data[7:9])
        hari_bulan = monthrange(tahun1, bulan1)[1]
        # Baca data
        x = [['AM' for u in range(24)] for u in range(hari_bulan)]
        y = [['AM' for u in range(24)] for u in range(hari_bulan)]
        z = [['AM' for u in range(24)] for u in range(hari_bulan)]
        f = [['AM' for u in range(24)] for u in range(hari_bulan)]
        h = [['AM' for u in range(24)] for u in range(hari_bulan)]
        d = [['AM' for u in range(24)] for u in range(hari_bulan)]
        i = [['AM' for u in range(24)] for u in range(hari_bulan)]
        
        # Set progressbar
        value = self.progressBar.value()
        self.progressBar.setMaximum(len(listfile))
        
        # Print keterangan
        self.labelfolder_3.setText('Creating excel file, please wait...')
        
        # Konversi data dari IAGA ke IMFV dengan fungsi iaga2imfv
        for j in range(0,len(listfile)):
            data = min2hour(listfile[j])
            tanggal = data[7]-1
            x[tanggal] = data[0]
            y[tanggal] = data[1]
            z[tanggal] = data[2]
            f[tanggal] = data[3]
            h[tanggal] = data[4]
            d[tanggal] = data[5]
            i[tanggal] = data[6]
            iaga2imfv(listfile[j],dirIAGA_)
            print(os.path.basename(glob.glob(os.path.join(str(dirIAGA_), '*.min'))[j]))
            value = j+1
            self.progressBar.setValue(value)    
            
        # Konversi ke format excel    
        format_excel(x,y,z,f,h,d,i,dirIAGA_)
        # Selesai
        self.labelfolder_3.setText('Done')
        
    # Fungsi untuk program stepremoval
    def stepremoval(self):
        # Ambil threshold dari form
        threshold = self.lineEdit.displayText()
        # Ambil tanggal dari widget calendar
        date = self.CalendarWidget.selectedDate().toPyDate()
        print('Removing steps at '+date.strftime('%d %B %Y'))
        # Jalankan program step removal dari folder berikut
        #step_removal(date,float(threshold),'Z:\\PROCESS\\DATALEMIRAWFILTER\\',home+'\\data\\Provisional')
        step_removal(date,float(threshold),home+'\\data\\LEMIRAW',home+'\\data\\Provisional')
    
    # Ambil variabel tanggal dari widget calendar
    def showDate(self):     
        date = self.CalendarWidget.selectedDate()
        #self.label.setText(str(date.toPyDate()))

# Jalankan program
if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)

    # Create and display the splash screen
    splash_pix = QtGui.QPixmap('splash.jpg')
    splash = QtWidgets.QSplashScreen(splash_pix, QtCore.Qt.WindowStaysOnTopHint)
    splash.setMask(splash_pix.mask())
    splash.show()
    app.processEvents()

    # Pause 2 detik
    time.sleep(2)

    # Program utama
    window = MyApp()
    window.show()
    splash.finish(window)
    sys.exit(app.exec_())

step_remove.py
# -*- coding: utf-8 -*-
"""
Created on Sun Dec 03 21:54:41 2017

@author: TUNTUNGAN
"""

from datetime import timedelta,datetime
import numpy as np
import re
import os.path
from scipy import signal
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates

def step_removal_process(data,threshold,comp):
    #Smooth the data using a median filter to remove spikes and the Savitzky-Golay running average filter
    Binterp = data.interpolate(method = 'linear')
    medfiltB = signal.medfilt(Binterp.values[:], 3);
    vAvgResampB = signal.savgol_filter(medfiltB,17,1);
    
    diffB = np.diff(vAvgResampB)
    
    steps = [i for i,v in enumerate(abs(diffB)) if v > threshold]

    if np.array(steps).size:
        breakpoints = [i for i,v in enumerate(np.diff(steps)) if v > 1]
        print('Number of ',comp,' breakpoints: '+str(len(breakpoints)))

        #Go through each break point and sum up the accumulated change
        breakpoints =  [-1]+breakpoints+[len(steps)-1]  # put a zero at the start for each of looping and the final point
        for i in range(0,len(breakpoints)-1):
            st = breakpoints[i]+1; en =  breakpoints[i+1]
            #Extend the edges of the step detection back and forward a few
            #points and check you haven't gone over the edge of the time-series
            stp=steps[st]-2 if steps[st]>3 else steps[st]
            enp=steps[en]+2 if steps[en]<(len(diffB)-3) else steps[en]
            #Compute the change in the field across the step by cumulatively
            #summing up the differences from point-to-point
            change = np.cumsum(diffB[stp:enp+1])
            #Now remove it from the time series starting at the revelant point
            vAvgResampB[stp:enp+1] = vAvgResampB[stp:enp+1] - change
            vAvgResampB[enp+1:-1] = vAvgResampB[enp+1:-1] - change[-1]
            
    return vAvgResampB
          
def step_removal(date1,threshold,namefolderlemi,namefolder_out):           
    #Define data
    tahun1 = int((date1-timedelta(days=1)).year)
    bulan1 = int((date1-timedelta(days=1)).month)
    tanggal1 = int((date1-timedelta(days=1)).day)
    tahun2 = int(date1.year)
    bulan2 = int(date1.month)
    tanggal2 = int(date1.day)
    hh=np.zeros(1440)
    dd=np.zeros(1440)
    zz=np.zeros(1440)
    ff=np.zeros(1440)
    xx=np.zeros(1440)
    yy=np.zeros(1440)
    h_mean=np.zeros(1440)
    d_mean=np.zeros(1440)
    z_mean=np.zeros(1440)
    f_mean=np.zeros(1440)
    x_mean=np.zeros(1440)
    y_mean=np.zeros(1440)
    
    #Koefisien Gaussian filter
    gaussFilter=[0.00045933,0.00054772,0.00065055,0.00076964,0.00090693,
        0.00106449,0.00124449,0.00144918,0.00168089,0.00194194,0.00223468,
        0.0025614,0.0029243,0.00332543,0.00376666,0.00424959,0.00477552,
        0.00534535,0.00595955,0.00661811,0.00732042,0.0080653,0.0088509,
        0.00967467,0.01053338,0.01142303,0.01233892,0.01327563,0.01422707,
        0.01518651,0.01614667,0.01709976,0.01803763,0.01895183,0.01983377,
        0.0206748,0.02146643,0.02220039,0.02286881,0.02346437,0.0239804,
        0.02441104,0.02475132,0.02499727,0.02514602,0.0251958,0.02514602,
        0.02499727,0.02475132,0.02441104,0.0239804,0.02346437,0.02286881,
        0.02220039,0.02146643,0.0206748,0.01983377,0.01895183,0.01803763,
        0.01709976,0.01614667,0.01518651,0.01422707,0.01327563,0.01233892,
        0.01142303,0.01053338,0.00967467,0.0088509,0.0080653,0.00732042,
        0.00661811,0.00595955,0.00534535,0.00477552,0.00424959,0.00376666,
        0.00332543,0.0029243,0.0025614,0.00223468,0.00194194,0.00168089,
        0.00144918,0.00124449,0.00106449,0.00090693,0.00076964,0.00065055,
        0.00054772,0.00045933]
    
    #Parameter Stasiun
    with open('station.ini') as f_init:
        content_init = f_init.readlines()
        datasta = list(np.tile(np.nan,len(content_init)))		
        for i in range(0,len(content_init)):
            datasta[i] = re.split('\=|\n',content_init[i])
                
    #Import data Lemi   
    filename1 = '%4.0f\\%02.0f\\%4.0f %02.0f %02.0f 00 00 00.txt' % (tahun1,bulan1,tahun1,bulan1,tanggal1)
    filename2 = '%4.0f\\%02.0f\\%4.0f %02.0f %02.0f 00 00 00.txt' % (tahun2,bulan2,tahun2,bulan2,tanggal2)
    
    try:
#    if os.path.isfile(namefolderlemi + '/' + filename1)==True and os.path.isfile(namefolderlemi + '/' + filename2)==True:
        with open(namefolderlemi + '/' + filename1) as f_lemi1:
            content1 = f_lemi1.readlines()
            dates1 = list(np.tile(np.nan,len(content1)))
            for i in range(0,len(content1)):
                data1 = re.split('\s+',content1[i])
                Bx1 = float(data1[6]) #+ 41300
                By1 = float(data1[7])
                Bz1 = float(data1[8]) #- 7100
                dates1[i] = list((data1[0]+'-'+data1[1]+'-'+data1[2]+' '+data1[3]+':'+data1[4]+':'+data1[5],Bx1,By1,Bz1))
            datadates1 = pd.DataFrame(dates1,columns=['dt', 'Bx', 'By', 'Bz'])
    except:
        print('File '+filename1+' not exist')
        
    try:
        with open(namefolderlemi + '/' + filename2) as f_lemi2:
            content2 = f_lemi2.readlines()
            dates2 = list(np.tile(np.nan,len(content2)))
            for i in range(0,len(content2)):
                data2 = re.split('\s+',content2[i])
                Bx2 = float(data2[6]) #+ 41300
                By2 = float(data2[7])
                Bz2 = float(data2[8]) #- 7100
                dates2[i] = list((data2[0]+'-'+data2[1]+'-'+data2[2]+' '+data2[3]+':'+data2[4]+':'+data2[5],Bx2,By2,Bz2))
            datadates2 = pd.DataFrame(dates2,columns=['dt', 'Bx', 'By', 'Bz'])
    except:
        print('File '+filename2+' not exist'    )
    
    timefull = pd.date_range(date1-timedelta(days=1), periods=172800, freq='s')
    timeadd = pd.concat([datadates1,datadates2], ignore_index=True)
    timeadd.index = pd.DatetimeIndex(timeadd.dt)
    datafull = timeadd.reindex(timefull)

    #Now look for steps and remove them ...
    #Threshold is how large in nT/sample do you want to consider a step? it is important to set this correctly

    #Removing step 
    vAvgResampX = step_removal_process(datafull.Bx,threshold,'X')
    vAvgResampY = step_removal_process(datafull.By,threshold,'Y')
    vAvgResampZ = step_removal_process(datafull.Bz,threshold,'Z')

    #Smooth the data using a median filter to remove final spikes 
    finalX = signal.medfilt(vAvgResampX, 3)
    finalY = signal.medfilt(vAvgResampY, 3)
    finalZ = signal.medfilt(vAvgResampZ, 3)  

    finalF = np.sqrt((finalX**2)+(finalY**2)+(finalZ**2))
    finalH = np.sqrt(finalX**2+finalY**2)
    finalD = 60*np.degrees(np.arcsin(finalY/finalH))   
    
    #1 Minute mean value using gaussian filter
    gaussFilter=gaussFilter/np.sum(gaussFilter)
    for i in range(0,1440):
        H_conv=np.convolve(finalH[86354+(60*i):86445+(60*i)],gaussFilter)
        D_conv=np.convolve(finalD[86354+(60*i):86445+(60*i)],gaussFilter)
        Z_conv=np.convolve(finalZ[86354+(60*i):86445+(60*i)],gaussFilter)
        F_conv=np.convolve(finalF[86354+(60*i):86445+(60*i)],gaussFilter)
        X_conv=np.convolve(finalX[86354+(60*i):86445+(60*i)],gaussFilter)
        Y_conv=np.convolve(finalY[86354+(60*i):86445+(60*i)],gaussFilter)
        hh[i]=H_conv[90]
        dd[i]=D_conv[90]
        zz[i]=Z_conv[90]
        ff[i]=F_conv[90]
        xx[i]=X_conv[90]
        yy[i]=Y_conv[90] 

    #Hitung rata-rata satu menit
    for k in range(0,1440):
        if np.isnan(hh[k])==True:
            h_mean[k]=99999.00
        else:
            h_mean[k]=hh[k]
        if np.isnan(dd[k])==True:
            d_mean[k]=99999.00
        else:
            d_mean[k]=dd[k]
        if np.isnan(zz[k])==True:
            z_mean[k]=99999.00
        else:
            z_mean[k]=zz[k]
        if np.isnan(ff[k])==True:
            f_mean[k]=99999.00
        else:
            f_mean[k]=ff[k]
        if np.isnan(xx[k])==True:
            x_mean[k]=99999.00
        else:
            x_mean[k]=xx[k]
        if np.isnan(yy[k])==True:
            y_mean[k]=99999.00
        else:
            y_mean[k]=yy[k]

    #Menyimpan file output
    fileout = 'TUN'+str(date1.year)+str(date1.month).zfill(2)+str(date1.day).zfill(2)+'pmin'+'.min'
    namefolder = namefolder_out
    f_iaga = open(namefolder + '/' + fileout, 'w')
    
    f_iaga.write(' Format                 IAGA-2002                                    |\n')
    f_iaga.write(' Source of Data         BMKG                                         |\n')
    f_iaga.write(' Station Name           %-45s|\n'%datasta[1][1])
    f_iaga.write(' IAGA Code              %-45s|\n'%datasta[2][1])
    f_iaga.write(' Geodetic Latitude      %-45s|\n'%datasta[8][1])
    f_iaga.write(' Geodetic Longitude     %-45s|\n'%datasta[9][1])
    f_iaga.write(' Elevation              %-45s|\n'%datasta[5][1])
    f_iaga.write(' Reported               DHZF                                         |\n')
    f_iaga.write(' Sensor Orientation     XYZ                                          |\n')
    f_iaga.write(' Digital Sampling       1 second                                     |\n')
    f_iaga.write(' Data Interval Type     Filtered 1-minute (00:15-01:45)              |\n')
    f_iaga.write(' Data Type              provisional                                  |\n')
    f_iaga.write(' #F=sqrt((Bx^2)+(By^2)+(Bz^2)),H=sqrt(Bx^2+By^2),D=60*asind(By/H)    |\n')
    f_iaga.write('DATE       TIME         DOY     TUND      TUNH      TUNZ      TUNF   |\n')
    
    for j in range(0,1440):
        body_iaga = '%s    %9.2f %9.2f %9.2f %9.2f\n' %((datetime(year=int(date1.year),month=int(date1.month),
                    day=int(date1.day))+timedelta(minutes=j)).strftime("%Y-%m-%d %H:%M:00.000 %j"),
                    d_mean[j],h_mean[j],z_mean[j],f_mean[j])
        f_iaga.write(body_iaga)
    f_iaga.close()
    
    print(fileout+' has been created')
    print('Creating graphic, please wait...')

    #PLOT GRAFIK
    t=pd.date_range(date1.strftime('%d-%B-%Y 00:00'),date1.strftime('%d-%B-%Y 23:59'),freq="1min")
    plt.rcParams['axes.linewidth'] = 0.3
    xfmt = mdates.DateFormatter('%H:%M')
    fig, axs = plt.subplots(3, 2, sharex=True)
    fig.subplots_adjust(hspace=0)
    fig.suptitle(str.upper(date1.strftime('%d %B %Y')), fontsize=16, fontweight='bold', color='k')
    fig.set_size_inches(12.5, 8.5, forward=True)
    
    #PLOT Sebelum
    dataX = datafull.Bx[86400:172800]
    dataY = datafull.By[86400:172800]
    dataZ = datafull.Bz[86400:172800]
    
    #X Sebelum
    axs[0,0].plot(dataX, linewidth=1, color='r')
    axs[0,0].grid(b=True, which='major', color='c', linestyle=':', linewidth=0.5)
    axs[0,0].set_ylabel('X COMPONENT\n (nT)', fontsize=10)
    axs[0,0].set_title('SEBELUM', fontsize=12, fontweight='bold', color='r')
    
    #Y Sebelum
    axs[1,0].plot(dataY, linewidth=1, color='g')
    axs[1,0].grid(b=True, which='major', color='c', linestyle=':', linewidth=0.5)
    axs[1,0].set_ylabel('Y COMPONENT\n (nT)', fontsize=10)
    
    #Z Sebelum
    axs[2,0].plot(dataZ, linewidth=1, color='b')
    axs[2,0].grid(b=True, which='major', color='c', linestyle=':', linewidth=0.5)
    axs[2,0].set_ylabel('Z COMPONENT\n (nT)', fontsize=10)
    axs[2,0].set_xlabel('Time', fontsize=10)
    
    
    #PLOT Sesudah
    #X Sesudah
    axs[0,1].plot(t, x_mean, linewidth=1, color='r')
    axs[0,1].grid(b=True, which='major', color='c', linestyle=':', linewidth=0.5)
    axs[0,1].set_title('SESUDAH', fontsize=12, fontweight='bold', color='r')
    axs[0,1].set_ylabel('X COMPONENT\n (nT)', fontsize=10)
    axs[0,1].yaxis.set_label_position("right")
    axs[0,1].yaxis.tick_right()
    
    #Y Sesudah
    axs[1,1].plot(t, y_mean, linewidth=1, color='g')
    axs[1,1].grid(b=True, which='major', color='c', linestyle=':', linewidth=0.5)
    axs[1,1].set_ylabel('Y COMPONENT\n (nT)', fontsize=10)
    axs[1,1].yaxis.set_label_position("right")
    axs[1,1].yaxis.tick_right()
    
    #Z Sesudah
    axs[2,1].plot(t, z_mean, linewidth=1, color='b')
    axs[2,1].grid(b=True, which='major', color='c', linestyle=':', linewidth=0.5)
    axs[2,1].set_ylabel('Z COMPONENT\n (nT)', fontsize=10)
    axs[2,1].yaxis.set_label_position("right")
    axs[2,1].yaxis.tick_right()
    axs[2,1].set_xlabel('Time', fontsize=10)
    axs[2,1].xaxis.set_major_formatter(xfmt)
    axs[2,1].set_xlim(min(t),max(t))
	
    plt.show()

    print('Finish\n')